api_gen.py

```diff
@@ -0,0 +1,124 @@
+It generates API and formats user and generated APIs.
+"""
+
+import os
+import shutil
+
+import namex
+
+package = "keras_nlp"
+
+
+def ignore_files(_, filenames):
+    return [f for f in filenames if f.endswith("_test.py")]
+
+
+def copy_source_to_build_directory(root_path):
+    # Copy sources (`keras_nlp/` directory and setup files) to build dir
+    build_dir = os.path.join(root_path, "tmp_build_dir")
+    if os.path.exists(build_dir):
+        shutil.rmtree(build_dir)
+    os.mkdir(build_dir)
+    shutil.copytree(
+        package, os.path.join(build_dir, package), ignore=ignore_files
+    )
+    return build_dir
+
+
+def export_version_string(api_init_fname):
+    with open(api_init_fname) as f:
+        contents = f.read()
+    with open(api_init_fname, "w") as f:
+        contents += "from keras_nlp.src.version_utils import __version__\n"
+        f.write(contents)
+
+
+def update_package_init(init_fname):
+    contents = """
+# Import everything from /api/ into keras.
+from keras_nlp.api import *  # noqa: F403
+from keras_nlp.api import __version__  # Import * ignores names start with "_".
+
+import os
+
+# Add everything in /api/ to the module search path.
+__path__.append(os.path.join(os.path.dirname(__file__), "api"))  # noqa: F405
+
+# Don't pollute namespace.
+del os
+
+# Never autocomplete `.src` or `.api` on an imported keras object.
+def __dir__():
+    keys = dict.fromkeys((globals().keys()))
+    keys.pop("src")
+    keys.pop("api")
+    return list(keys)
+
+
+# Don't import `.src` or `.api` during `from keras import *`.
+__all__ = [
+    name
+    for name in globals().keys()
+    if not (name.startswith("_") or name in ("src", "api"))
+]"""
+    with open(init_fname) as f:
+        init_contents = f.read()
+    with open(init_fname, "w") as f:
+        f.write(init_contents.replace("\nfrom keras_nlp import api", contents))
+
+
+def build():
+    # Backup the `keras_nlp/__init__.py` and restore it on error in api gen.
+    root_path = os.path.dirname(os.path.abspath(__file__))
+    code_api_dir = os.path.join(root_path, package, "api")
+    code_init_fname = os.path.join(root_path, package, "__init__.py")
+    # Create temp build dir
+    build_dir = copy_source_to_build_directory(root_path)
+    build_api_dir = os.path.join(build_dir, package, "api")
+    build_init_fname = os.path.join(build_dir, package, "__init__.py")
+    build_api_init_fname = os.path.join(build_api_dir, "__init__.py")
+    try:
+        os.chdir(build_dir)
+        # Generates `keras_nlp/api` directory.
+        if os.path.exists(build_api_dir):
+            shutil.rmtree(build_api_dir)
+        if os.path.exists(build_init_fname):
+            os.remove(build_init_fname)
+        os.makedirs(build_api_dir)
+        namex.generate_api_files(
+            "keras_nlp", code_directory="src", target_directory="api"
+        )
+        # Creates `keras_nlp/__init__.py` importing from `keras_nlp/api`
+        update_package_init(build_init_fname)
```

Comment:
It looks like this is to avoid the new toplevel `__init__` being generated by namex. Why not just skip `shutil.copy(build_init_fname, code_init_fname)` below and just let our toplevel `__init__` stay unaltered?

Would be nice to avoid the whole python file as a string thing we got going here, and it seems like there is no need for it. We aren't actually templating anything in.

Reply:
Totally makes sense. I made this change. Thanks for the feedback on it. It greatly simplifies it.